lib/kpeg/code_generator.rb

```diff
@@ -128,129 +119,102 @@ def output_op(code, op, indent=2)
             right = op.fin[0]
           end
 
-          code << indentify("  unless _tmp >= #{left} and _tmp <= #{right}\n", indent)
-          code << indentify("    self.pos = #{ss}\n", indent)
-          code << indentify("    _tmp = nil\n", indent)
-          code << indentify("  end\n", indent)
-          code << indentify("end\n", indent)
+          code << indentify("_tmp = get_byte(#{left}..#{right})\n", indent)
         else
           raise "Unsupported char range - #{op.inspect}"
         end
       when Choice
         ss = save()
         code << "\n"
-        code << indentify("#{ss} = self.pos\n", indent)
-        code << indentify("while true # choice\n", indent)
+        code << indentify("begin # choice\n", indent)
         op.ops.each_with_index do |n,idx|
-          output_op code, n, (indent+1)
-
-          code << indentify("  break if _tmp\n", indent)
-          code << indentify("  self.pos = #{ss}\n", indent)
-          if idx == op.ops.size - 1
-            code << indentify("  break\n", indent)
```

Comment:
Totally get the logic behind not restoring the save here, but we should double check that all the rules, if they consume input, perform the reset themselves. This was always a bit of a failsafe.

Reply:
I don't think double check pays the bill here. Rules are thoroughly tested by using them within compiller itself. "What could go wrong?" :-)

Reply:
Ha! fair enough. You're right through, the metacircularity is the best test bed.